packages/core/src/router/routing-table.ts

```diff
@@ -140,34 +139,21 @@ export class Route {
     const pathParams = this._buildPathParams(match);
     debug(' -> found with params: %j', pathParams);
 
-    return this._createResolvedRoute(pathParams);
-  }
-
-  private _createResolvedRoute(pathParams: PathParameterValues): ResolvedRoute {
-    return this._handler ?
-      this._createResolvedHandlerRoute(pathParams) :
-      this._createResolvedControllerRoute(pathParams);
-  }
-
-  private _createResolvedHandlerRoute(
-    pathParams: PathParameterValues,
-  ): ResolvedHandlerRoute {
     return {
-      handler: this._handler!,
-      spec: this.spec,
+      route: this,
       pathParams: pathParams,
     };
   }
 
-  private _createResolvedControllerRoute(
-    pathParams: PathParameterValues,
-  ): ResolvedControllerRoute {
-    return {
-      controllerName: this._controllerName!,
-      methodName: this.spec['x-operation-name']!,
-      spec: this.spec,
-      pathParams: pathParams,
-    };
+  abstract updateBindings(requestContext: Context): void;
+
+  abstract invokeHandler(
+    requestContext: Context,
+    args: OperationArgs,
+  ): Promise<OperationRetval>;
```

Comment:
A couple of ideas to consider: 

Should we always pass `requestContext` as the first argument of the route handler function? It would allow handler functions to access the per-request context. 

```ts
app.route(new Route('get', '/greet', spec, async (ctx, name) => {
  const req = await ctx.get('http.request');
  const host = req.headers['host'];
  return `hello ${world} from ${host}`;
});
```

Can we think of a better alternative for context propagation when defining routes as functions? Also, should we pass `requestContext` to the custom sequence handler function registered via `app.handler()`?

```ts
app.handler((sequence, ctx, req, res) => {
  // ...
});
```

Let's keep these proposed changes out of scope of this pull request though. If you think these two ideas are worth discussing further, then let's open a new issue.